### PR TITLE
Support Proxy Configuration from Environment Variables also refactor existing ProxyConfig

### PR DESCRIPTION
**Pull request by joviegas**
<!--- Provide a general summary of your changes in the Title above -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
## Support Proxy Configuration from Environment Variables also refactor existing ProxyConfig
### Case 1 When both  System  and Environment Variable are set

```
// System Properties 
"http.proxyHost", SYSTEM_PROPERTY_HOST.com
"http.proxyPort", 2222
"http.nonProxyHosts", SYSTEM_PROPERTY_NON_PROXY
"http.password", SYSTEM_PROPERTY_PASSWORD
"http.user", SYSTEM_PROPERTY_USER

// Environment Variables Properties 
"HTTP_PROXY" : http://EnvironmentUser:EnvironmentPassword@ENVIRONMENT_VARIABLE_HOST:3333
"NO_PROXY" : "environmentnonproxy.host"
```


| Builder                                                                                                                                                                                                                                | Resolved Proxy                                                                                                                                                       |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <br>ProxyConfiguration._builder_()<br>.build();                                                                                                                                                                                        | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY |
| <br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.build();                                                                                                                                    | Host =localhost<br>Port = 1234<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY                |
| <br><br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.password("password")<br>.username("userName")<br>.nonProxyHosts(Stream._of_("userNonProxy").collect(Collectors._toSet_()))<br>.build(); | Host =localhost<br>Port = 1234<br>Password = password<br>UserName = userName<br>Non ProxyHost = userNonProxy                                                         |
| <br>ProxyConfiguration._builder_()<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                                                                       | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY |
| <br>ProxyConfiguration._builder_()<br>.useSystemPropertyValues(Boolean._FALSE_)<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                          | Host =ENVIRONMENT_VARIABLE_HOST.com<br>Port = 3333<br>Password = EnvironmentPassword<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host       |


### Case 2: No  System Property Setting, but Environment Variables Are Set

```
// System Properties 
none

// Environment Variables Properties 
"HTTP_PROXY" : http://EnvironmentUser:EnvironmentPassword@ENVIRONMENT_VARIABLE_HOST:3333
"NO_PROXY" : "environmentnonproxy.host"
```
| Builder                                                                                                                                                                                                                                | Resolved Proxy                                                                                                                                                 |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <br>ProxyConfiguration._builder_()<br>.build();                                                                                                                                                                                        | Host =ENVIRONMENT_VARIABLE_HOST.com<br>Port = 3333<br>Password = EnvironmentPassword<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host |
| <br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.build();                                                                                                                                    | Host =localhost<br>Port = 1234<br>Password = EnvironmentPassword<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host                     |
| <br><br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.password("password")<br>.username("userName")<br>.nonProxyHosts(Stream._of_("userNonProxy").collect(Collectors._toSet_()))<br>.build(); | Host =localhost<br>Port = 1234<br>Password = password<br>UserName = userName<br>Non ProxyHost = userNonProxy                                                   |
| <br>ProxyConfiguration._builder_()<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                                                                       | Host =ENVIRONMENT_VARIABLE_HOST.com<br>Port = 3333<br>Password = EnvironmentPassword<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host |
| <br>ProxyConfiguration._builder_()<br>.useSystemPropertyValues(Boolean._TRUE_)<br>.build();                                                                                                                                            | Host =ENVIRONMENT_VARIABLE_HOST.com<br>Port = 3333<br>Password = EnvironmentPassword<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host |
| <br>ProxyConfiguration._builder_()<br>.useEnvironmentVariableValues(Boolean.FALSE)<br>.build();          

### Case 3: With  System Property Setting, but No Environment Variables
```
// System Properties 
"http.proxyHost", SYSTEM_PROPERTY_HOST.com
"http.proxyPort", 2222
"http.nonProxyHosts", SYSTEM_PROPERTY_NON_PROXY
"http.password", SYSTEM_PROPERTY_PASSWORD
"http.user", SYSTEM_PROPERTY_USER

// Environment Variables Properties
None
```
| Builder                                                                                                                                                                                                                                | Resolved Proxy                                                                                                                                                       |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <br>ProxyConfiguration._builder_()<br>.build();                                                                                                                                                                                        | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY |
| <br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.build();                                                                                                                                    | Host =localhost<br>Port = 1234<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY                |
| <br><br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.password("password")<br>.username("userName")<br>.nonProxyHosts(Stream._of_("userNonProxy").collect(Collectors._toSet_()))<br>.build(); | Host =localhost<br>Port = 1234<br>Password = password<br>UserName = userName<br>Non ProxyHost = userNonProxy                                                         |
| <br>ProxyConfiguration._builder_()<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                                                                       | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = SYSTEM_PROPERTY_USER<br>Non ProxyHost = SYSTEM_PROPERTY_NON_PROXY |
| <br>ProxyConfiguration._builder_()<br>.useSystemPropertyValues(Boolean._FALSE_)<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                          | Host =Null<br>Port = 0<br>Password = Null<br>UserName = Null<br>Non ProxyHost = Null                                                                                 |


### Case 4 When No  System  and No Environment Variable are set
```
// System Properties 
None

// Environment Variables Properties 
None
```
| Builder                                                                                                                                                                                                                                | Resolved Proxy                                                                                               |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
| <br>ProxyConfiguration._builder_()<br>.build();                                                                                                                                                                                        | Host =Null<br>Port = 0<br>Password = Null<br>UserName = Null<br>Non ProxyHost = Null                         |
| <br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.build();                                                                                                                                    | Host =localhost<br>Port = 1234<br>Password = null<br>UserName = null<br>Non ProxyHost = null                 |
| <br><br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.password("password")<br>.username("userName")<br>.nonProxyHosts(Stream._of_("userNonProxy").collect(Collectors._toSet_()))<br>.build(); | Host =localhost<br>Port = 1234<br>Password = password<br>UserName = userName<br>Non ProxyHost = userNonProxy |
| <br>ProxyConfiguration._builder_()<br>.useSystemPropertyValues(Boolean._FALSE_) // Or Null Or True<br>.useEnvironmentVariableValues(Boolean._TRUE_) // Or Null Or False<br>.build();                                                   | Host =Null<br>Port = 0<br>Password = Null<br>UserName = Null<br>Non ProxyHost = Null                         |

### Case 5 Mix and Match of UserInfo
```
// System Properties 
// There is no User name in system property
// No Non proxy host here
"http.proxyHost", SYSTEM_PROPERTY_HOST.com
"http.proxyPort", 2222
"http.password", SYSTEM_PROPERTY_PASSWORD


// Environment Variables Properties 
"HTTP_PROXY" : http://EnvironmentUser@ENVIRONMENT_VARIABLE_HOST:3333
"NO_PROXY" : "environmentnonproxy.host"
```
| Builder                                                                                                                                                                                                                                | Resolved Proxy                                                                                                                                                      |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <br>ProxyConfiguration._builder_()<br>.build();                                                                                                                                                                                        | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = null // (confirm !!)<br>Non ProxyHost = environmentnonproxy.host |
| <br>ProxyConfiguration._builder_()<br>.password("fromBuilderPassword")<br>.build();                                                                                                                                                    | Host =localhost<br>Port = 2222<br>Password = fromBuilderPassword<br>UserName = null<br>Non ProxyHost = environmentnonproxy.host                                     |
| <br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.build();                                                                                                                                    | Host =localhost<br>Port = 1234<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = null<br>Non ProxyHost = environmentnonproxy.host                                |
| <br><br>ProxyConfiguration._builder_()<br>.endpoint(URI._create_("http://localhost:1234"))<br>.password("password")<br>.username("userName")<br>.nonProxyHosts(Stream._of_("userNonProxy").collect(Collectors._toSet_()))<br>.build(); | Host =localhost<br>Port = 1234<br>Password = password<br>UserName = userName<br>Non ProxyHost = userNonProxy                                                        |
| <br>ProxyConfiguration._builder_()<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                                                                       | Host =SYSTEM_PROPERTY_HOST.com<br>Port = 2222<br>Password = SYSTEM_PROPERTY_PASSWORD<br>UserName = null<br>Non ProxyHost = environmentnonproxy.host                 |
| <br>ProxyConfiguration._builder_()<br>.useSystemPropertyValues(Boolean._FALSE_)<br>.useEnvironmentVariableValues(Boolean._TRUE_)<br>.build();                                                                                          | Host =ENVIRONMENT_VARIABLE_HOST.com<br>Port = 3333<br>Password = null<br>UserName = EnvironmentUser<br>Non ProxyHost = environmentnonproxy.host                     |


## Modifications
<!--- Describe your changes in detail -->
- Refactored Proxy settings to pick the Local settings like System and Environment settings
- Modified Apache, Crt, Netty and Url http client to resolve Proxy config based on above settings

## Testing
<!--- Please describe in detail how you tested your changes -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Added Junits

## Screenshots (if appropriate)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## License
<!--- The SDK is released under the Apache 2.0 license (http://aws.amazon.com/apache2.0/), so any code you submit will be released under that license -->
<!--- For substantial contributions, we may ask you to sign a Contributor License Agreement (http://en.wikipedia.org/wiki/Contributor_License_Agreement) -->
<!--- Put an `x` in the below box if you confirm that this request can be released under the Apache 2 license -->
- [X] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by dagnir** [comment]
UserSetValuesOnSdkBuilder | useSystemProperty | useEnvironmentVariable | ResolvedProxy
-- | -- | -- | --
Null | true | (But ENV_VAR is SET) | Environment Variables

What does this mean exactly? Are system properties set in this case? If so, shouldn't precedence be given to system properties?

File: utils/src/main/java/software/amazon/awssdk/utils/internal/proxy/ProxyEnvironmentVariableConfigProvider.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+
+    private Optional<URL> silentlyGetUrl() {
+        String stringUrl = Objects.equals(this.scheme, HTTPS) ? ProxyEnvironmentSetting.HTTPS_PROXY.getStringValue().orElse(null)
+                                                              : ProxyEnvironmentSetting.HTTP_PROXY.getStringValue().orElse(null);
+        if (StringUtils.isNotBlank(stringUrl)) {
+            try {
+                return Optional.of(new URL(stringUrl));
+            } catch (MalformedURLException e) {
+                log.error(() -> "Malformed environment  url string " + stringUrl, e);
```
Comment (dagnir) [inline utils/src/main/java/software/amazon/awssdk/utils/internal/proxy/ProxyEnvironmentVariableConfigProvider.java:60]: nit: extra space before 'url'

also Can we say "Malformed proxy environment variable" or something like that so it's clearer when reading the logs?

File: test/http-client-tests/src/main/java/software/amazon/awssdk/http/proxy/ProxyConfigCommonTestData.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    private static TestProxySetting getSystemPropertyProxySettings() {
+        return new TestProxySetting().host(SYSTEM_PROPERTY_HOST)
+                                     .port(Integer.parseInt(SYSTEM_PROPERTY_PORT_NUMBER))
+                                     .userName(SYSTEM_PROPERTY_USER)
+                                     .password(SYSTEM_PROPERTY_PASSWORD)
+                                     .nonProxyHost(SYSTEM_PROPERTY_NON_PROXY);
+    }
+
+
+    private static TestProxySetting getEnvironmentalVariableProxySettings() {
```
[the remark this answers is not in the document]
Comment (joviegas) [reply test/http-client-tests/src/main/java/software/amazon/awssdk/http/proxy/ProxyConfigCommonTestData.java:282]: 🌲🌲🌲🌲🌲🌲

File: http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/ProxyConfiguration.java [hunk 3]
```diff
@@ -46,19 +47,40 @@ public final class ProxyConfiguration implements ToCopyableBuilder<ProxyConfigur
...
+
+        this.username = builder.username != null || proxyConfigProvider == null ? builder.username :
+                        proxyConfigProvider.userName().orElseGet(() -> builder.username);
+
+        this.password = builder.password != null || proxyConfigProvider == null ? builder.password :
+                        proxyConfigProvider.password().orElseGet(() -> builder.password);
+
+        this.nonProxyHosts = builder.nonProxyHosts != null || proxyConfigProvider == null ? builder.nonProxyHosts :
+                             proxyConfigProvider.nonProxyHosts();
+
```
[the remark this answers is not in the document]
Comment (joviegas) [reply http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/ProxyConfiguration.java:73]: done

File: utils/src/main/java/software/amazon/awssdk/utils/ProxyEnvironmentSetting.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    ProxyEnvironmentSetting(String environmentVariable) {
+        this.environmentVariable = environmentVariable;
+    }
+
+    @Override
+    public Optional<String> getStringValue() {
+        Optional<String> envVarLowercase = resolveEnvironmentVariable(environmentVariable);
+        if (envVarLowercase.isPresent() && !envVarLowercase.get().trim().isEmpty()) {
+            return envVarLowercase.map(String::trim);
```
[the remark this answers is not in the document]
Comment (joviegas) [reply utils/src/main/java/software/amazon/awssdk/utils/ProxyEnvironmentSetting.java:46]: done

File: utils/src/main/java/software/amazon/awssdk/utils/internal/proxy/ProxyEnvironmentVariableConfigProvider.java (hunk 1 above)
Comment (joviegas) [reply utils/src/main/java/software/amazon/awssdk/utils/internal/proxy/ProxyEnvironmentVariableConfigProvider.java:60]: done

**Comment by joviegas** [comment]
> UserSetValuesOnSdkBuilder 	useSystemProperty 	useEnvironmentVariable 	ResolvedProxy
> Null 	true 	(But ENV_VAR is SET) 	Environment Variables
> 
> What does this mean exactly? Are system properties set in this case? If so, shouldn't precedence be given to system properties?

Updated the table to be more clear , same table which was used for internal discussion

**Review by dagnir — APPROVED** [review]

**Pull request merged**